File: src/ApocalypseIndicators.php
<?php
/**
 * NWO Signal Spectrum - Apocalypse Indicators Integration
 * 
 * Adds apocalyptic signal types to the existing Signal Spectrum system
 * Sources: Aircraft, seismic, solar, radiation, asteroids, pandemics
 */

namespace NWOSignalSpectrum;

class ApocalypseIndicators {
    
    private $db;
    private $cache;
    
    // API Endpoints
    const USGS_EARTHQUAKE_API = 'https://earthquake.usgs.gov/fdsnws/event/1/query';
    const NASA_NEO_API = 'https://api.nasa.gov/neo/rest/v1/feed';
    const NOAA_SPACE_WEATHER = 'https://services.swpc.noaa.gov/products';
    const SAFECAST_API = 'https://api.safecast.org/measurements.json';
    const ADSB_EXCHANGE = 'https://api.adsbexchange.com/v2';
    const OPEN_METEO = 'https://api.open-meteo.com/v1';
    
    // Anomaly thresholds
    const THRESHOLDS = [
        'aircraft_spike' => 3.0,      // 3x normal activity
        'earthquake_cluster' => 3,     // 3+ M6+ in 24h
        'solar_flare_x_class' => 1,    // Any X-class flare
        'radiation_spike' => 10.0,     // 10x baseline
        'asteroid_close_approach' => 5, // <5 lunar distances
        'geomagnetic_storm' => 7,      // Kp index >= 7
    ];
    
    public function __construct($database, $cache = null) {
        $this->db = $database;
        $this->cache = $cache;
    }
    
    /**
     * Detect aviation anomalies (rich people fleeing)
     * Tracks private jet spikes from specific locations
     */
    public function detectAviationAnomaly($region = null) {
        // Query ADS-B data for business jet activity
        $params = [
            'format' => 'json',
            'filter' => 'bizjets', // Filter for business jets
        ];
        
        if ($region) {
            $params['bounds'] = $this->getRegionBounds($region);
        }
        
        $data = $this->fetch(self::ADSB_EXCHANGE . '/feed', $params);
        
        if (!$data || !isset($data['ac'])) {
            return null;
        }
        
        $currentCount = count($data['ac']);
        $baseline = $this->getHistoricalBaseline('aviation', $region, 168); // 7 days
        
        if ($baseline == 0) $baseline = 1;
        
        $anomalyScore = $currentCount / $baseline;
        
        if ($anomalyScore >= self::THRESHOLDS['aircraft_spike']) {
            return [
                'type' => 'aviation_anomaly',
                'severity' => $this->calculateSeverity($anomalyScore, self::THRESHOLDS['aircraft_spike']),
                'current_count' => $currentCount,
                'baseline' => $baseline,
                'anomaly_score' => $anomalyScore,
                'region' => $region,
                'timestamp' => time(),
                'description' => sprintf(
                    'Unusual spike in business jet activity: %d aircraft (%.1fx normal)',
                    $currentCount,
                    $anomalyScore
                )
            ];
        }
        
        return null;
    }
    
    /**
     * Monitor seismic activity for earthquake clusters
     */
    public function detectSeismicAnomaly($hours = 24) {
        $params = [
            'format' => 'geojson',
            'starttime' => date('Y-m-d\TH:i:s', strtotime("-{$hours} hours")),
            'minmagnitude' => 6.0,
            'orderby' => 'time',
        ];
        
        $data = $this->fetch(self::USGS_EARTHQUAKE_API, $params);
        
        if (!$data || !isset($data['features'])) {
            return null;
        }
        
        $significantQuakes = $data['features'];
        $count = count($significantQuakes);
        
        // Check for clustering (multiple large quakes in short timeframe)
        if ($count >= self::THRESHOLDS['earthquake_cluster']) {
            $maxMagnitude = 0;
            $locations = [];
            
            foreach ($significantQuakes as $quake) {
                $mag = $quake['properties']['mag'];
                $place = $quake['properties']['place'];
                
                if ($mag > $maxMagnitude) $maxMagnitude = $mag;
                $locations[] = $place;
            }
            
            return [
                'type' => 'seismic_cluster',
                'severity' => $maxMagnitude >= 7.0 ? 'critical' : 'high',
                'quake_count' => $count,
                'max_magnitude' => $maxMagnitude,
                'locations' => array_unique($locations),
                'timestamp' => time(),
                'description' => sprintf(
                    'Cluster of %d significant earthquakes detected (max M%.1f)',
                    $count,
                    $maxMagnitude
                )
            ];
        }
        
        return null;
    }
    
    /**
     * Monitor solar activity for dangerous flares/CMEs
     */
    public function detectSolarAnomaly() {
        // Get latest space weather data from NOAA
        $data = $this->fetch(self::NOAA_SPACE_WEATHER . '/noaa-planetary-k-index.json');
        $kpIndex = $this->getCurrentKpIndex($data);
        
        // Get X-ray flux for flare detection
        $xrayData = $this->fetch(self::NOAA_SPACE_WEATHER . '/goes-xray-flux-primary.json');
        $flareClass = $this->getFlareClass($xrayData);
        
        $anomalies = [];
        
        // Check for geomagnetic storm
        if ($kpIndex >= self::THRESHOLDS['geomagnetic_storm']) {
            $anomalies[] = [
                'type' => 'geomagnetic_storm',
                'severity' => $kpIndex >= 9 ? 'extreme' : ($kpIndex >= 8 ? 'severe' : 'strong'),
                'kp_index' => $kpIndex,
                'timestamp' => time(),
                'description' => "Geomagnetic storm in progress (Kp {$kpIndex})"
            ];
        }
        
        // Check for X-class solar flare
        if ($flareClass === 'X') {
            $anomalies[] = [
                'type' => 'solar_flare',
                'severity' => 'critical',
                'flare_class' => 'X',
                'timestamp' => time(),
                'description' => 'X-class solar flare detected - potential CME'
            ];
        }
        
        return count($anomalies) > 0 ? $anomalies : null;
    }
    
    /**
     * Monitor radiation levels globally
     */
    public function detectRadiationAnomaly() {
        // Get recent Safecast measurements
        $params = [
            'capture' => 'none',
            'per_page' => 1000,
            'since' => date('c', strtotime('-1 hour')),
        ];
        
        $data = $this->fetch(self::SAFECAST_API, $params);
        
        if (!$data || !isset($data['measurements'])) {
            return null;
        }
        
        $spikes = [];
        
        foreach ($data['measurements'] as $reading) {
            $value = $reading['value']; // μSv/h
            $location = $reading['location_name'] ?? 'Unknown';
            $baseline = $this->getRadiationBaseline($location);
            
            if ($baseline > 0 && ($value / $baseline) >= self::THRESHOLDS['radiation_spike']) {
                $spikes[] = [
                    'location' => $location,
                    'value' => $value,
                    'baseline' => $baseline,
                    'multiplier' => $value / $baseline
                ];
            }
        }
        
        if (count($spikes) >= 3) { // Multiple locations spiking
            return [
                'type' => 'radiation_spike',
                'severity' => 'critical',
                'spike_count' => count($spikes),
                'locations' => array_column($spikes, 'location'),
                'timestamp' => time(),
                'description' => sprintf(
                    'Radiation spikes detected at %d locations',
                    count($spikes)
                )
            ];
        }
        
        return null;
    }
    
    /**
     * Monitor near-Earth objects for close approaches
     */
    public function detectAsteroidThreat($days = 7) {
        $apiKey = getenv('NASA_API_KEY') ?: 'DEMO_KEY';
        
        $params = [
            'start_date' => date('Y-m-d'),
            'end_date' => date('Y-m-d', strtotime("+{$days} days")),
            'api_key' => $apiKey,
        ];
        
        $data = $this->fetch(self::NASA_NEO_API, $params);
        
        if (!$data || !isset($data['near_earth_objects'])) {
            return null;
        }
        
        $threats = [];
        $lunarDistance = 384400; // km
        
        foreach ($data['near_earth_objects'] as $date => $objects) {
            foreach ($objects as $neo) {
                $distance = $neo['close_approach_data'][0]['miss_distance']['kilometers'] ?? null;
                $diameter = $neo['estimated_diameter']['meters']['estimated_diameter_max'] ?? 0;
                $isHazardous = $neo['is_potentially_hazardous_asteroid'] ?? false;
                
                if ($distance && ($distance / $lunarDistance) <= self::THRESHOLDS['asteroid_close_approach']) {
                    if ($isHazardous || $diameter > 100) {
                        $threats[] = [
                            'name' => $neo['name'],
                            'diameter_m' => $diameter,
                            'distance_km' => (float)$distance,
                            'distance_ld' => (float)$distance / $lunarDistance,
                            'approach_date' => $date,
                            'is_hazardous' => $isHazardous
                        ];
                    }
                }
            }
        }
        
        if (count($threats) > 0) {
            usort($threats, function($a, $b) {
                return $a['distance_ld'] <=> $b['distance_ld'];
            });
            
            $closest = $threats[0];
            
            return [
                'type' => 'asteroid_close_approach',
                'severity' => $closest['is_hazardous'] ? 'critical' : 'high',
                'object_count' => count($threats),
                'closest_object' => $closest['name'],
                'closest_distance_ld' => round($closest['distance_ld'], 2),
                'diameter_m' => $closest['diameter_m'],
                'approach_date' => $closest['approach_date'],
                'timestamp' => time(),
                'description' => sprintf(
                    '%s approaching within %.1f lunar distances (%.0f m diameter)',
                    $closest['name'],
                    $closest['distance_ld'],
                    $closest['diameter_m']
                )
            ];
        }
        
        return null;
    }
    
    /**
     * Run all apocalypse indicators
     */
    public function runAllChecks() {
        $alerts = [];
        
        // Run all detection methods
        $checks = [
            'aviation' => $this->detectAviationAnomaly(),
            'seismic' => $this->detectSeismicAnomaly(),
            'solar' => $this->detectSolarAnomaly(),
            'radiation' => $this->detectRadiationAnomaly(),
            'asteroid' => $this->detectAsteroidThreat(),
        ];
        
        foreach ($checks as $type => $result) {
            if ($result) {
                if (is_array($result) && isset($result[0])) {
                    // Multiple anomalies from one check
                    foreach ($result as $anomaly) {
                        $alerts[] = $anomaly;
                        $this->storeSignal($anomaly);
                    }
                } else {
                    $alerts[] = $result;
                    $this->storeSignal($result);
                }
            }
        }
        
        return $alerts;
    }
    
    /**
     * Store detected signal in database
     */
    private function storeSignal($signal) {
        $stmt = $this->db->prepare("
            INSERT INTO apocalypse_signals 
            (type, severity, description, metadata, created_at)
            VALUES (?, ?, ?, ?, NOW())
        ");
        
        $metadata = json_encode([
            'current_count' => $signal['current_count'] ?? null,
            'baseline' => $signal['baseline'] ?? null,
            'anomaly_score' => $signal['anomaly_score'] ?? null,
            'max_magnitude' => $signal['max_magnitude'] ?? null,
            'kp_index' => $signal['kp_index'] ?? null,
            'distance_ld' => $signal['closest_distance_ld'] ?? null,
        ]);
        
        $stmt->execute([
            $signal['type'],
            $signal['severity'],
            $signal['description'],
            $metadata
        ]);
    }
    
    /**
     * Calculate apocalypse level (1-5) based on active signals
     */
    public function calculateApocalypseLevel() {
        $stmt = $this->db->query("
            SELECT severity, COUNT(*) as count 
            FROM apocalypse_signals 
            WHERE created_at > DATE_SUB(NOW(), INTERVAL 24 HOUR)
            GROUP BY severity
        ");
        
        $scores = [
            'low' => 1,
            'medium' => 2,
            'high' => 3,
            'critical' => 4,
            'extreme' => 5
        ];
        
        $totalScore = 0;
        $maxSeverity = 0;
        
        while ($row = $stmt->fetch()) {
            $score = $scores[$row['severity']] ?? 1;
            $totalScore += $score * $row['count'];
            if ($score > $maxSeverity) $maxSeverity = $score;
        }
        
        // Level is max of: highest severity OR cumulative score threshold
        $level = max($maxSeverity, min(5, ceil($totalScore / 5)));
        
        return [
            'level' => $level,
            'description' => $this->getLevelDescription($level),
            'active_signals' => $totalScore,
            'timestamp' => time()
        ];
    }
    
    private function getLevelDescription($level) {
        $descriptions = [
            1 => 'Normal - No significant anomalies detected',
            2 => 'Elevated - Minor anomalies present',
            3 => 'High - Multiple concerning signals',
            4 => 'Severe - Critical conditions developing',
            5 => 'Extreme - Apocalyptic conditions possible'
        ];
        return $descriptions[$level] ?? 'Unknown';
    }
    
    // Helper methods
    private function fetch($url, $params = []) {
        if (!empty($params)) {
            $url .= '?' . http_build_query($params);
        }
        
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
        
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        
        if ($httpCode !== 200 || !$response) {
            return null;
        }
        
        return json_decode($response, true);
    }
    
    private function getHistoricalBaseline($type, $region, $hours) {
        // Query database for historical average
        $stmt = $this->db->prepare("
            SELECT AVG(value) as baseline 
            FROM signal_baselines 
            WHERE type = ? AND region = ? AND created_at > DATE_SUB(NOW(), INTERVAL ? HOUR)
        ");
        $stmt->execute([$type, $region, $hours]);
        $result = $stmt->fetch();
        return $result['baseline'] ?? 0;
    }
    
    private function getRadiationBaseline($location) {
        // Default baseline radiation ~0.1 μSv/h
        return 0.1;
    }
    
    private function getRegionBounds($region) {
        // Return lat/lon bounds for regions
        $bounds = [
            'davos' => [46.8, 47.0, 9.8, 10.0],
            'maui' => [20.5, 21.0, -156.5, -155.5],
            'monaco' => [43.7, 43.8, 7.4, 7.5],
        ];
        return $bounds[$region] ?? null;
    }
    
    private function calculateSeverity($score, $threshold) {
        $ratio = $score / $threshold;
        if ($ratio >= 10) return 'extreme';
        if ($ratio >= 5) return 'critical';
        if ($ratio >= 3) return 'high';
        if ($ratio >= 2) return 'medium';
        return 'low';
    }
    
    private function getCurrentKpIndex($data) {
        if (!$data || empty($data)) return 0;
        $latest = end($data);
        return isset($latest[1]) ? (int)$latest[1] : 0;
    }
    
    private function getFlareClass($data) {
        if (!$data || empty($data)) return null;
        // Parse X-ray flux and return flare class (A, B, C, M, X)
        $latest = end($data);
        $flux = isset($latest[1]) ? (float)$latest[1] : 0;
        
        if ($flux >= 1e-4) return 'X';
        if ($flux >= 1e-5) return 'M';
        if ($flux >= 1e-6) return 'C';
        if ($flux >= 1e-7) return 'B';
        return 'A';
    }
}

File: app/services/apocalypse_indicators.py
"""Apocalypse indicator service.

Direct port of the PHP ApocalypseIndicators class — all six detection categories
(aviation, seismic, solar, radiation, asteroid, RF) plus the rollup that produces
a level 1–5 from the active signals.
"""
from __future__ import annotations

import logging
from datetime import datetime, timedelta, timezone
from typing import Any

import asyncpg
import httpx

from app.config import get_settings

logger = logging.getLogger(__name__)


LEVEL_LABELS = {
    1: "Normal — baseline activity",
    2: "Elevated — minor anomalies",
    3: "High — multiple concerning signals",
    4: "Severe — coordinated indicators",
    5: "Extreme — multi-source critical alerts",
}


# -------- Level computation --------

async def current_level(conn: asyncpg.Connection) -> dict[str, Any]:
    """Aggregate active signals in the last 24h into a single level."""
    rows = await conn.fetch(
        """
        SELECT type, severity, COUNT(*)::int AS n
        FROM apocalypse_signals
        WHERE created_at > NOW() - INTERVAL '24 hours'
        GROUP BY type, severity
        """
    )

    breakdown: dict[str, int] = {}
    severity_weights = {"low": 1, "medium": 2, "high": 3, "critical": 4, "extreme": 5}
    total_weight = 0
    total_signals = 0

    for r in rows:
        breakdown[r["type"]] = breakdown.get(r["type"], 0) + r["n"]
        total_weight += severity_weights.get(r["severity"], 1) * r["n"]
        total_signals += r["n"]

    if total_signals == 0:
        level = 1
    else:
        avg_severity = total_weight / total_signals
        # Bin avg severity 1.0–5.0 into discrete levels.
        level = max(1, min(5, round(avg_severity)))

    return {
        "level": level,
        "description": LEVEL_LABELS[level],
        "active_signals": total_signals,
        "breakdown": breakdown,
        "timestamp": datetime.now(timezone.utc),
    }


async def record_level(conn: asyncpg.Connection, level: int, active: int) -> None:
    """Snapshot the current level into the history table."""
    await conn.execute(
        """
        INSERT INTO apocalypse_level_history (level, description, active_signals)
        VALUES ($1, $2, $3)
        """,
        level,
        LEVEL_LABELS.get(level, ""),
        active,
    )


async def level_history(
    conn: asyncpg.Connection, hours: int = 24
) -> list[dict[str, Any]]:
    rows = await conn.fetch(
        """
        SELECT level, description, active_signals, recorded_at
        FROM apocalypse_level_history
        WHERE recorded_at > NOW() - ($1 || ' hours')::INTERVAL
        ORDER BY recorded_at DESC
        """,
        str(hours),
    )
    return [dict(r) for r in rows]


# -------- Aviation (ADS-B Exchange) --------

async def detect_aviation_anomaly(
    conn: asyncpg.Connection, region: str = "global"
) -> dict[str, Any] | None:
    """Compare current business-jet count in the region to baseline. >2σ → anomaly."""
    settings = get_settings()
    baseline_row = await conn.fetchrow(
        "SELECT value FROM signal_baselines WHERE type='aviation' AND region=$1", region
    )
    baseline = float(baseline_row["value"]) if baseline_row else 100.0

    # Pull recent count from the local cache table; cron job updates it from ADS-B.
    row = await conn.fetchrow(
        """
        SELECT COUNT(*)::int AS n
        FROM aircraft_sightings
        WHERE is_business_jet = TRUE
          AND detected_at > NOW() - INTERVAL '1 hour'
        """
    )
    current = int(row["n"]) if row else 0

    if baseline <= 0:
        return None
    ratio = current / baseline
    if ratio < 2.0:
        return None

    return {
        "type": "aviation",
        "severity": "critical" if ratio > 4 else "high" if ratio > 3 else "medium",
        "description": (
            f"{current} business jets in {region} — {ratio:.1f}x baseline ({baseline:.0f})"
        ),
        "metadata": {"current": current, "baseline": baseline, "ratio": ratio},
        "region": region,
    }


# -------- Seismic (USGS) --------

async def detect_seismic_anomaly(
    conn: asyncpg.Connection, hours: int = 24, magnitude_min: float = 6.0
) -> dict[str, Any] | None:
    row = await conn.fetchrow(
        """
        SELECT COUNT(*)::int AS n, MAX(magnitude) AS max_mag
        FROM seismic_events
        WHERE event_time > NOW() - ($1 || ' hours')::INTERVAL
          AND magnitude >= $2
        """,
        str(hours),
        magnitude_min,
    )
    n = int(row["n"]) if row else 0
    if n < 3:
        return None
    max_mag = float(row["max_mag"]) if row and row["max_mag"] is not None else 0.0
    return {
        "type": "seismic",
        "severity": "critical" if max_mag >= 7 else "high",
        "description": f"Cluster of {n} M{magnitude_min}+ earthquakes in {hours}h (max M{max_mag:.1f})",
        "metadata": {"count": n, "max_magnitude": max_mag, "hours": hours},
        "region": "global",
    }


# -------- Solar (NOAA SWPC) --------

async def fetch_noaa_solar() -> dict[str, Any] | None:
    """Pull current solar activity. Cached by caller if needed."""
    settings = get_settings()
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(f"{settings.noaa_swpc_url}/goes/primary/xrays-1-day.json")
            r.raise_for_status()
            data = r.json()
        # Find the highest flare class in the last 24h
        flux_max = max((float(d.get("flux", 0) or 0) for d in data), default=0.0)
        if flux_max >= 1e-4:
            klass = "X"
        elif flux_max >= 1e-5:
            klass = "M"
        elif flux_max >= 1e-6:
            klass = "C"
        else:
            klass = None
        return {"flux_max": flux_max, "class": klass}
    except httpx.HTTPError as exc:
        logger.warning("NOAA solar fetch failed: %s", exc)
        return None


async def detect_solar_anomaly() -> dict[str, Any] | None:
    solar = await fetch_noaa_solar()
    if not solar or not solar.get("class"):
        return None
    klass = solar["class"]
    if klass not in ("M", "X"):
        return None
    return {
        "type": "solar",
        "severity": "extreme" if klass == "X" else "high",
        "description": f"{klass}-class solar flare detected",
        "metadata": solar,
        "region": "global",
    }


# -------- Radiation (Safecast) --------

async def detect_radiation_anomaly(
    conn: asyncpg.Connection,
) -> dict[str, Any] | None:
    row = await conn.fetchrow(
        """
        SELECT COUNT(*)::int AS n, MAX(deviation_percent) AS max_dev
        FROM radiation_readings
        WHERE measured_at > NOW() - INTERVAL '6 hours'
          AND deviation_percent > 200
        """
    )
    n = int(row["n"]) if row else 0
    if n < 1:
        return None
    return {
        "type": "radiation",
        "severity": "critical",
        "description": f"{n} radiation sensors >2x baseline in last 6h",
        "metadata": {"count": n, "max_deviation_pct": float(row["max_dev"] or 0)},
        "region": "global",
    }


# -------- Asteroids (NASA NEO) --------

async def detect_asteroid_anomaly(conn: asyncpg.Connection) -> dict[str, Any] | None:
    row = await conn.fetchrow(
        """
        SELECT name, miss_distance_ld, diameter_max_m
        FROM neo_objects
        WHERE approach_date BETWEEN CURRENT_DATE AND CURRENT_DATE + INTERVAL '7 days'
          AND is_hazardous = TRUE
        ORDER BY miss_distance_ld ASC
        LIMIT 1
        """
    )
    if not row:
        return None
    return {
        "type": "asteroid",
        "severity": "high" if float(row["miss_distance_ld"]) < 1.0 else "medium",
        "description": (
            f"Hazardous NEO {row['name']} approaching at "
            f"{float(row['miss_distance_ld']):.2f} lunar distances"
        ),
        "metadata": {
            "name": row["name"],
            "diameter_m": float(row["diameter_max_m"] or 0),
            "miss_distance_ld": float(row["miss_distance_ld"]),
        },
        "region": "space",
    }


# -------- Composite check (cron-driven) --------

async def run_all_checks(conn: asyncpg.Connection) -> list[dict[str, Any]]:
    """Run every detector. Any non-None result is persisted as a new signal."""
    checks = [
        detect_aviation_anomaly(conn),
        detect_seismic_anomaly(conn),
        detect_solar_anomaly(),
        detect_radiation_anomaly(conn),
        detect_asteroid_anomaly(conn),
    ]
    import asyncio

    results = await asyncio.gather(*checks, return_exceptions=True)
    fresh: list[dict[str, Any]] = []
    for r in results:
        if isinstance(r, Exception):
            logger.exception("detector error: %s", r)
            continue
        if r is None:
            continue
        await conn.execute(
            """
            INSERT INTO apocalypse_signals (type, severity, description, metadata, region)
            VALUES ($1, $2, $3, $4::jsonb, $5)
            """,
            r["type"],
            r["severity"],
            r["description"],
            __import__("json").dumps(r.get("metadata") or {}),
            r.get("region"),
        )
        fresh.append(r)
    return fresh
